grafana/grafana
# Dashboards: Block updates on soft deleted dashboard

**What is this feature?**

It fixes a special case when restoring a dashboard. It's alternative to #94674

**Why do we need this feature?**

It fixes a bug when saving a dashboard via the `POST /api/dashboards/db` API using a UID of a soft-delete dashboard:
the requested used to succeed but the dashboard was not visible because it was marked as deleted. This fix blocks such request; hence, the users should hard delete the dashboard first using: `DELETE /dashboards/uid/{uid}/trash` and then try to save the dashboard.

**Who is this feature for?**

This bug was causing an issue for Cloud provider app to re-install dashboards if the dashboards/enclosing folder
was deleted while the `dashboardRestore` was set 

**Which issue(s) does this PR fix?**:



Fixes #

**Special notes for your reviewer:**
Steps to test it in cloud:
- navigate to the [ephemeral instance](https://ephemeral1511182194720papagia.grafana-dev.net/)
- install the cloud provider dashboards
- get the UIDs of the created dashboards
- uninstall them
- re-installing attempt will fail with: `A soft deleted dashboard with the same uid already exists`
- hard delete the dashboards using by `DELETE /dashboards/uid/{uid}/trash` or from the UI
- re-installing attempt will fail succeed

Steps to test it locally:
- enable the feature: `GF_FEATURE_TOGGLES_ENABLE=dashboardRestore`
- create a dashboard
- note down its UID
- delete the dashboard
- try to save a dashboard using the specific UID (that should fail):

<details><summary>example request</summary>
<p>

```
curl -X 'POST' \
  'http://<user>:<password>@localhost:3000/api/dashboards/db' \
  -H 'accept: application/json' \
  -H 'Content-Type: application/json' \
  -d '{
  "UpdatedAt": "2024-10-17T07:58:15.844Z",
  "dashboard": {
    "title": "some other",
    "uid": "fe14j23396n7kd"
  },
  "isFolder": false,
  "overwrite": true
}'
{"message":"A soft deleted dashboard with the same uid already exists"}%
```

</p>
</details> 

- hard delete the dashboard using: `DELETE /dashboards/uid/{uid}/trash`
- try the above request (it should succeed this time)

Please check that:
